server/src/main/java/org/elasticsearch/cluster/NodeConnectionsService.java

```diff
@@ -347,4 +357,113 @@ public String toString() {
             }
         }
     }
+
+    private class ConnectionHistory {
```

Comment:
Yeah I like the look of this. Maybe `ConnectionHistory implements TransportConnectionListener` rather than having another layer of indirection?

Also this needs to be covered in `NodeConnectionsServiceTests`.